spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTransactionManager.java

```diff
@@ -0,0 +1,240 @@
+/*
+ *  Copyright 2018 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.data.datastore.core;
+
+import java.util.Iterator;
+import java.util.List;
+
+import com.google.cloud.datastore.Datastore;
+import com.google.cloud.datastore.DatastoreException;
+import com.google.cloud.datastore.Entity;
+import com.google.cloud.datastore.FullEntity;
+import com.google.cloud.datastore.Key;
+import com.google.cloud.datastore.Query;
+import com.google.cloud.datastore.QueryResults;
+import com.google.cloud.datastore.Transaction;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.protobuf.ByteString;
+
+import org.springframework.transaction.NoTransactionException;
+import org.springframework.transaction.TransactionDefinition;
+import org.springframework.transaction.TransactionException;
+import org.springframework.transaction.TransactionSystemException;
+import org.springframework.transaction.interceptor.TransactionAspectSupport;
+import org.springframework.transaction.support.AbstractPlatformTransactionManager;
+import org.springframework.transaction.support.DefaultTransactionStatus;
+
+/**
+ * Cloud Datastore transaction manager
+ *
+ * @author Chengyuan Zhao
+ *
+ * @since 1.1
+ */
+public class DatastoreTransactionManager extends AbstractPlatformTransactionManager {
+	private final Datastore datastore;
+
+	public DatastoreTransactionManager(final Datastore datastore) {
+		this.datastore = datastore;
+	}
+
+	@VisibleForTesting
+	Tx getCurrentTX() {
+		try {
+			return (Tx) ((DefaultTransactionStatus) TransactionAspectSupport
+					.currentTransactionStatus()).getTransaction();
+		}
+		catch (NoTransactionException e) {
+			return null;
+		}
+	}
+
+	@Override
+	protected Object doGetTransaction() throws TransactionException {
+		Tx tx = getCurrentTX();
+		if (tx != null && tx.transaction != null && tx.transaction.isActive()) {
+			return tx;
+		}
+		tx = new Tx();
+		return tx;
+	}
+
+	@Override
+	protected void doBegin(Object transactionObject,
+			TransactionDefinition transactionDefinition) throws TransactionException {
+		if (transactionDefinition
+				.getIsolationLevel() != TransactionDefinition.ISOLATION_DEFAULT) {
+			throw new IllegalStateException(
+					"SpannerTransactionManager supports only isolation level TransactionDefinition.ISOLATION_DEFAULT");
```

Comment:
Spanner -> Datastore

Reply:
Check if any other isolation levels are supported

Reply:
look into "serializable"

Reply:
Reading through the available isolation levels in `TransactionDefinition`, it appears `ISOLATION_DEFAULT or ISOLATION_SERIALIZABLE` can be supported.